contributing/architecture-eval-triggers.md

```diff
@@ -0,0 +1,274 @@
+  any follow-up Evaluations associated with scheduling, planning, or
+  deployments.
+* **periodic-job**: A periodic job that hits its timer and dispatches a child
+  job will result in 1 Evaluation created, plus any additional Evaluations
+  associated with scheduling or planning.
+* **node-drain**: Draining a node will create 1 Evaluation for each Job on the
+  node that's draining, plus any additional Evaluations associated with
+  scheduling or planning.
+* **node-update**: When the fingerprint of a client node has changed or the node
+  has changed state (from up to down), Nomad creates 1 Evaluation for each Job
+  running on the Node, plus 1 Evaluation for each system job that has
+  `datacenters` that include the datacenter for that Node.
+* **alloc-stop**: When the API that serves the `nomad alloc stop` command is
+  hit, Nomad creates 1 Evaluation.
+* **scheduled**: Nomad's internal housekeeping will periodically create
+  Evaluations for garbage collection.
+* **rolling-update**: When a `system` job is updated, the [`update`][] block's
+  `stagger` field controls how many Allocations will be scheduled at a time. The
+  scheduler will create 1 Evaluation to follow-up for the next set.
+* **deployment-watcher**: When a `service` job is updated, the [`update`][]
+  block controls how many Allocations will be scheduled at a time. The
+  deployment watcher runs on the leader and monitors Allocation healthy. It will
+  create 1 Evaluation when the Deployment has reached the next step.
+* **failed-follow-up**: Evaluations that hit a delivery limit and will not be
+  retried by the eval broker are marked as failed. The leader periodically
+  reaps failed Evaluations and creates 1 new Evaluation for these, with a delay.
+* **max-plan-attempts**: The scheduler will retry Evaluations that are rejected
+  by the plan applier with a new cluster state snapshot. If the scheduler
+  exceeds the maximum number of retries, ti will create 1 new Evaluation in the
+  `blocked` state.
+* **alloc-failure**: If an Allocation fails and exceeds its maximum
+  [`restart` attempts][], Nomad creates 1 new Evaluation.
+* **queued-allocs**: When a scheduler processes an Evaluation, it may not be
+  able to place all Allocations. It will create 1 new Evaluation in the
+  `blocked` state to be processed later when node updates arrive.
+* **preemption**: When Allocations are preempted, the plan applier creates 1
+  Evaluation for each Job that has been preempted.
+* **job-scaling**: Scaling a Job will result in 1 Evaluation created, plus any
+  follow-up Evaluations associated with scheduling, planning, or deployments.
+* **max-disconnect-timeout**: When an Allocation is in the `unknown` state for
+  longer than the [`max_client_disconnect`][] window, the scheduler will create
+  1 Evaluation.
+* **reconnect**: When a Node in the `disconnected` state reconnects, Nomad will
+  create 1 Evaluation.
```

Comment:
```suggestion
* **reconnect**: When a Node in the `disconnected` state reconnects, Nomad will
  create 1 Evaluation per job with an allocation on the reconnected Node.
```

I think? We actually create 1 eval per *allocation* but the batching in `node_endpoint.go` dedupes evals by job.

Interesting side note: when a down down reconnects I think it will call Node.Register and Node.UpdateAlloc in nondeterministic order, and if they occur within 50ms of each other the node endpoints batching will drop subsequent evals for a job from whichever RPC was processed last. I think this means when reconnecting an alloc you may get the following evals:

1. TriggeredBy: node-update (Register called first, UpdateAlloc called within 50ms later)
2. TriggeredBy: reconnect (UpdateAlloc called first, Register called within 50ms later)
3. Both node-update and reconnect evals because they occurred more than 50ms apart (or at least were not in the same batch)